specs/jsonschema-core.md

```diff
@@ -2029,32 +2029,129 @@ SHOULD use the terms defined by this document to do so.
-implementation can display them to end-users in violation of a spec, or fail to
-strip them if such behavior is expected.
-
-A malicious schema author could place executable code or other dangerous
-material within a `$comment`. Implementations MUST NOT parse or otherwise take
-action based on `$comment` contents.
+While schemas and instances are not always represented as JSON text, they are
+defined in terms of the JSON data model. As such, the security considerations
+defined in [RFC 8259][rfc8259] may still apply in environments where text-based
+representations are used, particularly those considerations related to parsing,
+number precision, and structural limitations.
+
+Schemas and instances are frequently authored by untrusted parties.
+Implementations that accept or evaluate such inputs may be exposed to several
+classes of attack, particularly denial-of-service (DoS) by means of resource
+exhaustion.
+
+### Nested `anyOf`/`oneOf`
+
+One risk for resource exhaustion in JSON Schema arises from the nested use of
+`anyOf` and `oneOf`. While a single combinator keyword with multiple subschemas
+is typically manageable, nesting them causes the number of evaluation paths to
+grow exponentially.
+
+For example, a `oneOf` with 5 subschemas, each containing another `oneOf` with 5
+options, results in 25 evaluation paths. Adding a third level increases this to
+125, and so on. Attackers can exploit this by crafting schemas that force
+validators to explore a large number of branches.
+
+This evaluation explosion is particularly dangerous when each path involves
+expensive work such as collecting large annotations or evaluating complex
+regular expressions. These effects multiply across paths and can result in
+excessive CPU or memory consumption, leading to denial-of-service.
+
+Implementations that evaluate untrusted schema are encouraged to take steps to
+mitigate these threats with measures such as bounding combinator keyword depth
+and breadth, limiting memory used for annotation collection, and guarding
+against resource-intensive validations such as pathological regexes.
+
+### Dynamic References
+
+The paper ["The Complexity of JSON Schema: Undecidable, Expensive, Yet
+Tractable" (Caroni et al., 2024)](https://doi.org/10.1145/3632891) has shown
+that validation in the presence of dynamic references is PSPACE-complete. The
+paper describes a method for replacing dynamic references with static ones, but
+doing so can cause the size of the schema to grow exponentially. Implementations
+should be aware of this risk and may wish to implement the method described in
+the paper or impose limits on dynamic reference resolution.
+
+### Infinite Loops and Cycles
+
+Infinite loops can occur when evaluating schemas that produce cycles during
+reference resolution. These cycles may involve multiple schemas. Not all
+recursive schemas create loops, but implementations are advised to detect and
+break these cycles when they are encountered.
```

Comment:
Earlier in the spec, implementations are _required_ to halt when encountering infinite cycles.

How would one break such a cycle?

Reply:
I didn't intend to imply that one could somehow continue evaluating the schema. I'll reword to make it more clear that it should stop evaluation when it detects a cycle.